server/src/main/java/org/elasticsearch/cluster/metadata/IndexReshardingState.java

```diff
@@ -0,0 +1,383 @@
+                this.code = code;
+            }
+
+            @Override
+            public void writeTo(StreamOutput out) throws IOException {
+                out.writeByte(this.code);
+            }
+
+            public static TargetShardState readFrom(StreamInput in) throws IOException {
+                var code = in.readByte();
+                return switch (code) {
+                    case 0 -> CLONE;
+                    case 1 -> HANDOFF;
+                    case 2 -> SPLIT;
+                    case 3 -> DONE;
+                    default -> throw new IllegalStateException("unknown target shard state [" + code + "]");
+                };
+            }
+        }
+
+        private static final ParseField SOURCE_SHARDS_FIELD = new ParseField("source_shards");
+        private static final ParseField TARGET_SHARDS_FIELD = new ParseField("target_shards");
+
+        @SuppressWarnings("unchecked")
+        private static final ConstructingObjectParser<Split, Void> SPLIT_PARSER = new ConstructingObjectParser<>(
+            "split",
+            args -> new Split(
+                ((List<SourceShardState>) args[0]).toArray(new SourceShardState[0]),
+                ((List<TargetShardState>) args[1]).toArray(new TargetShardState[0])
+            )
+        );
+
+        static {
+            SPLIT_PARSER.declareObjectArray(
+                ConstructingObjectParser.constructorArg(),
+                (parser, c) -> SourceShardState.valueOf(parser.text()),
+                SOURCE_SHARDS_FIELD
+            );
+            SPLIT_PARSER.declareObjectArray(
+                ConstructingObjectParser.constructorArg(),
+                (parser, c) -> TargetShardState.valueOf(parser.text()),
+                TARGET_SHARDS_FIELD
+            );
+        }
+
+        private final int oldShardCount;
+        private final int newShardCount;
+        private final SourceShardState[] sourceShards;
+        private final TargetShardState[] targetShards;
+
+        Split(SourceShardState[] sourceShards, TargetShardState[] targetShards) {
+            this.sourceShards = sourceShards;
+            this.targetShards = targetShards;
+
+            oldShardCount = sourceShards.length;
+            newShardCount = oldShardCount + targetShards.length;
+        }
+
+        Split(StreamInput in) throws IOException {
+            this(
+                in.readArray(SourceShardState::readFrom, SourceShardState[]::new),
+                in.readArray(TargetShardState::readFrom, TargetShardState[]::new)
+            );
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeArray(sourceShards);
+            out.writeArray(targetShards);
+        }
+
+        static Split fromXContent(XContentParser parser) throws IOException {
+            return SPLIT_PARSER.parse(parser, null);
+        }
+
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
```

Comment:
Override?

Reply:
Thanks, done in 20264b8. I'm surprised intellij didn't give me a squiggle for that one.